[Build Speed] Break up DocumentInlines.h

jernoble opened the pull request:
#### d84e3011acfc3494de30727755fbcf19cea2e995
<pre>
[Build Speed] Break up DocumentInlines.h
<a href="https://rdar.apple.com/161694748">rdar://161694748</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299920">https://bugs.webkit.org/show_bug.cgi?id=299920</a>

Reviewed by Ryosuke Niwa.

Prior to this patch, DocumentInlines.h was the most expensive header used in
WebCore; it is included 303 times in a WebCore Unified build, and on this
machine, each include of that file too the compile on average 1.5s to parse, for
a total of 8m CPU minutes of time spent parsing this header.

To diagnose why this header is so expensive, I commented out all the
implementations from the body of this header and ran a compile, collecting a
list of all the errors encountered. I then collected the counts of each error
message, and grouped liked errors together. It became clear that most uses of
DocumentInlines.h were to pull in a very narrow set of inlined methods, but
to get those few methods had to include a very large and very heavy header.

To address this, DocumentInlines.h was broken up into separate headers, each
which includes only Document.h and one or two other headers. Usually the second
header was required to define the return type of the method, and the type
definition is required because the return type is a CheckedPtr or RefPtr.
However because the return value is typically used at the call site, that return
value header would have been included already, meaning this style of header has
zero effective additional cost.

The files were created in descending order of the number of uses:
- DocumentView.h - 187
- DocumentPage.h - 175
- DocumentQuirks.h - 84
- DocumentSecurityOrigin.h - 54
- DocumentResourceLoader.h - 49
- DocumentMarkers.h - 27
- DocumentEventLoop.h - 26
- DocumentSettingsValues.h - 23
- DocumentWindow.h - 19

The remaining grab-bag of inlined methods had less than 10 uses each and were
left in DocumentInlines.h.

After this patch, DocumentInlines.h is used in just over 100 implementation
files and no headers. None of the new Document*.h headers or DocumentInlines.h
appear in the top 100 most expensive header files. DocumentInlines.h dropped
from the #1 most expensive header taking 8m to parse to #389 and 5.7s of
parsing. The most expensive new header is DocumentPage.h at #124 and 25s of
parsing.

Canonical link: <a href="https://commits.webkit.org/300989@main">https://commits.webkit.org/300989@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48a5034958a421572e65f9cae3a061b875225318

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124500 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44178 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34912 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131338 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76480 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3769f0db-cc96-4903-b6b7-e698a9dd5f1a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52759 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94711 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62814 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c7a4f192-8559-43c0-af09-854a556fd249) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127454 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111353 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75288 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34722 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29509 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74817 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134003 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51371 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39197 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103189 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102974 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26231 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48335 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26603 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48314 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51241 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57031 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50652 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54011 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52332 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/daa9c620eff01283c93bcc0bf054c89e4187a4e6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a5a10704e91e7791d07ed4a8c6e8489b2f1a5f56. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/217214db4eb0c7f79a721456478ec2c1bb068447. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

jernoble commented on Source/WebCore/dom/ElementInlines.h:
```diff
@@ -97,6 +97,20 @@ inline String Element::attributeTrimmedWithDefaultARIA(const QualifiedName& name
     return value.trim(isASCIIWhitespace).simplifyWhiteSpace(isASCIIWhitespace);
 }
 
+inline bool Document::shouldMaskURLForBindings(const URL& urlToMask) const
```
Because `ElementInlines.h` needs `Document::maskedURLForBindingsIfNeeded(...)`, which itself needs `Document::shouldMaskURLForBindings(...)`, in order to implement `Element::getURLAttributeForBindings(...)`. One reason `DocumentInlines.h` is included so much is because it's included by other `*Inlines.h` headers. And these `Document` methods are only called from `Element.cpp` and `HTMLImageElement.cpp`. Moving these implementations here means a `ElementInlines.h` doesn't need to include `DocumentInlines.h`, and causes significant reduction in the number of compilation units that include `DocumentInlines.h` for no reason.

In a future patch, I'll break up other `*Inlines.h` headers in the same way, and these might move to some other header.

jernoble replied:
We can absolutely try that; my goal was to not un-inline anything unless I could prove that inlining was useless somehow, like [private inlines](https://github.com/WebKit/WebKit/wiki/Analyzing-Build-Performance#avoid-private-inlines) or [virtual inlines](https://github.com/WebKit/WebKit/wiki/Analyzing-Build-Performance#avoid-virtual-inlines). But once this big patch lands, un-inlining would be a good next step.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/506a785b9af757e2bd6d1e04288893f7cd7e8d23. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6ba4ab0254d0eae77787a51e0c099494994f978e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0177a315d1900753885653083e3a3b5fa302edf4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/680fea4a8e2bcf53fecf57a10088225c1414075f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51638

jernoble committed:
[Build Speed] Break up DocumentInlines.h
rdar://161694748
https://bugs.webkit.org/show_bug.cgi?id=299920

Reviewed by Ryosuke Niwa.

Prior to this patch, DocumentInlines.h was the most expensive header used in
WebCore; it is included 303 times in a WebCore Unified build, and on this
machine, each include of that file too the compile on average 1.5s to parse, for
a total of 8m CPU minutes of time spent parsing this header.

To diagnose why this header is so expensive, I commented out all the
implementations from the body of this header and ran a compile, collecting a
list of all the errors encountered. I then collected the counts of each error
message, and grouped liked errors together. It became clear that most uses of
DocumentInlines.h were to pull in a very narrow set of inlined methods, but
to get those few methods had to include a very large and very heavy header.

To address this, DocumentInlines.h was broken up into separate headers, each
which includes only Document.h and one or two other headers. Usually the second
header was required to define the return type of the method, and the type
definition is required because the return type is a CheckedPtr or RefPtr.
However because the return value is typically used at the call site, that return
value header would have been included already, meaning this style of header has
zero effective additional cost.

The files were created in descending order of the number of uses:
- DocumentView.h - 187
- DocumentPage.h - 175
- DocumentQuirks.h - 84
- DocumentSecurityOrigin.h - 54
- DocumentResourceLoader.h - 49
- DocumentMarkers.h - 27
- DocumentEventLoop.h - 26
- DocumentSettingsValues.h - 23
- DocumentWindow.h - 19

The remaining grab-bag of inlined methods had less than 10 uses each and were
left in DocumentInlines.h.

After this patch, DocumentInlines.h is used in just over 100 implementation
files and no headers. None of the new Document*.h headers or DocumentInlines.h
appear in the top 100 most expensive header files. DocumentInlines.h dropped
from the #1 most expensive header taking 8m to parse to #389 and 5.7s of
parsing. The most expensive new header is DocumentPage.h at #124 and 25s of
parsing.

Canonical link: https://commits.webkit.org/300989@main
